ruamel-examples/parse_it.py

```diff
@@ -0,0 +1,237 @@
+        provenance : tuple
+            provenance[0]: line number
+            provenance[1]: column number
+        """
+
+        data = super().construct_object(node, *args, **kwargs)
+
+        provenance = (
+            node.start_mark.line,
+            node.start_mark.column,
+        )
+
+        return (data, provenance)
+
+
+class EnvironmentRepresenter(ruamel.yaml.RoundTripRepresenter):
+    """When dumping, this class is used to remove the !ENV tag from a particular value."""
+
+    def represent_scalar(self, tag, value, style=None, anchor=None):
+        type_tags = {
+            "int": f"tag:{YAML_SPEC_TAG}:int",
+            "float": f"tag:{YAML_SPEC_TAG}:float",
+            "bool": f"tag:{YAML_SPEC_TAG}:bool",
+            "str": f"tag:{YAML_SPEC_TAG}:str",
+        }
+        if tag == "!ENV":
+            type_tag = type(value).__name__
+            actual_type = type_tags[type_tag]
+            return super().represent_scalar(actual_type, value, style, anchor=anchor)
+        return super().represent_scalar(tag, value, style, anchor=anchor)
+
+
+class EsmToolsLoader(ruamel.yaml.YAML):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.filename = None
+        self.add_comments = True
+        self.Constructor = ProvenanceConstructor
+        self.Constructor.add_constructor(None, ProvenanceConstructor.construct_scalar)
+        self.Constructor.add_constructor(
+            "tag:yaml.org,2002:bool", ProvenanceConstructor.construct_yaml_bool
+        )
+        self.Representer = EnvironmentRepresenter
+
+    def get_filename(self):
+        return self.filename
+
+    def set_filename(self, filename):
+        self.filename = filename
+
+    def load(self, stream):
+        self.set_filename(stream.name)
+        mapping_with_tuple_prov = super().load(stream)
+
+
+        def _extract_dict(tup, config=None, prov=False):
+            """
+            Iterative function to extract dictionary with or without provenance
+
+            Parameters
+            ----------
+            tup : tuple
+            config: dictionary holding either main config of provenance
+            prov: If main or provenance dict should be returned (True or False, default False)
+
+            Returns
+            -------
+            canfig: dictionary with or wihtout provenance
+
+            """
+            config = config if config else {}
+# unpack tuple
+            diction, (line,col) =  tup
+# check if diction is ruamel.yaml.comments.CommentedMap (of class dict)
+            if isinstance(diction, ruamel.yaml.comments.CommentedMap):
+                for (key,prov1),(value,prov2) in diction.items():
+                    if isinstance(value, ruamel.yaml.comments.CommentedMap):
+                        config[key] = _extract_dict((value,prov2), config, prov)
+                    elif isinstance(value, list):
+                        templist = []
```

Comment:
instead of this you can have here the `data` and `provenance` dictionaries initialized:
```python3
data[key] = []
provenance[key] = []
```
Then you don't need the line at the end